approximate_string_matching/matching_with_dont_cares_and_mismatches.py

```diff
@@ -0,0 +1,284 @@
+  for mismatch_pos in mismatch_positions_in_text:
+    if mismatch_pos is not None and \
+       mismatch_pos < len(text_numeric) and \
+       mismatch_pos-i < len(word_numeric):
+
+      tv = text_numeric[mismatch_pos]
+      wv = word_numeric[mismatch_pos-i]
+      correlation_correction += tv*wv*(tv-wv)**2
+
+  return correlation_correction == clifford_array[i]
+
+
+def naive_hamming_with_mismatches(t, w, n, m, k):
+  for i in range(1, n - m + 2):
+    if distance.hamming_distance('#' + t[i:i + m], w, m, m, '?') <= k:
+      yield i
+
+
+# Strongly selective families
+# https://doi.org/10.48550/arXiv.0712.3876
+
+def generate_strongly_selective_family(n, r):
+  if r == 0:
+    return [[]]
+
+  if n == 1 or r*r*np.log(n) >= n:
+    return [{i} for i in range(n)]
+
+  delta = (r-1)/r
+  q = 2*r
+  while not is_prime(q): # after O(ln q) steps we'll find prime
+    q += 1
+
+  k = int(np.ceil(np.log(n) / np.log(q)))
+  m = int(np.ceil(k/(1-entropy(delta,q))))
+  matrix = linear_code_matrix_with_high_rate(m, k, delta, q)
+  S = {(i,l): set() for i in range(m) for l in range(q)}
+
+  for i,y in enumerate(itertools.product(range(q), repeat=k)):
+    z = (matrix@y)%q
+    for p in range(m):
+      if i < n:
+        S[(p, z[p])].add(i)
+
+  return list(x for x in S.values() if len(x) > 0)
+
+def linear_code_matrix_with_high_rate(m, k, delta, q):
+  G = np.zeros((m,k), dtype=int)
+  modular_inverse = all_modular_inverses_array(q)
+  number_of_non_zeroes_in_Gy = {
+    tuple(x_r)+(x,):0
+    for x in range(1,q)
+    for j in range(k)
+    for x_r in itertools.product(range(q), repeat=j)
+  }
+
+  for i in range(m):
+    for j in range(k):
+      W = np.zeros(q, dtype=np.float64)
+      for x in range(1,q):
+
+        # The original paper uses Gray code order to achieve constant
+        # time updates per changing x_r, but for this problem we can
+        # simply iterate in the default order and increase complexity
+        # by factor `k`, since constructing ssf takes less time than
+        # running the pattern matching algorithm (k in this algorithm
+        # is log m in the pattern matching problem).
+        for x_r in itertools.product(range(q), repeat=j):
+          v = (-modular_inverse[x])*sum(G[i][t]*x_r[t] for t in range(j)) % q
+          c = number_of_non_zeroes_in_Gy[tuple(x_r) + (x,)]
+          # Above line does the same as the commented code listed below,
+          # but has a better complexity, since it uses some memoization
+          #
+          # y = np.array(list(x_r)+[x]+[0]*(k-j-1))
+          # Gy = G@y
+          # c2 = sum(Gy[t]%q != 0 for t in range(i))
+          # assert c == c2
+
+          W[v] -= binom.pmf(np.floor(delta*m-c), m-j, 1-1/q)
+
+      G[i][j] = np.argmax(W)
+
+    # after selecting whole row G[i] update number of non zeroes
+    for j in range(k):
+      for x in range(1,q):
+        for x_r in itertools.product(range(q), repeat=j):
+          number_of_non_zeroes_in_Gy[tuple(x_r)+(x,)] += (
```

Comment:
I guess this is slightly cleaner visually:
if np.dot(G[i][:j+1], list(x_r)+[x])%q == 0:
  number_of_non_zeroes_in_Gy[tuple(x_r)+(x,)] += 1

Reply:
Sure, actually it should be if (...) != 0